crates/crabe_math/src/shape/line.rs

```diff
@@ -12,3 +12,148 @@ pub struct Line {
+
+        let denominator = ((self.start.x - self.end.x) * (line.start.y - line.end.y))
+            - ((self.start.y - self.end.y) * (line.start.x - line.end.x));
+        if denominator == 0. {
+            return None;
+        }
+
+        let t = t_nominator / denominator;
+        let u = u_nominator / denominator;
+        if t > 1. || t < 0. || u > 1. || u < 0. {
+            return None;
+        }
+        return Some(self.point_allong_line(t));
+    }
+
+    // return the closest point on the line
+    pub fn closest_point_on_line(&self, point: &Point2<f64>) -> Point2<f64> {
+        let line_direction = self.end - self.start;
+        let point_direction = *point - self.start;
+
+        let line_length_squared = line_direction.norm_squared();
+        if line_length_squared == 0.0 {
+            // The line segment has zero length, return the start point.
+            return self.start;
+        }
+        let t = point_direction.dot(&line_direction) / line_length_squared;
+
+        // The point is closest to a point on the segment.
+        self.start + t * line_direction
+    }
+
+    // return the closest point on the segment
+    pub fn closest_point_on_segment(&self, point: &Point2<f64>) -> Point2<f64> {
+        let line_direction = self.end - self.start;
+        let point_direction = *point - self.start;
+
+        let line_length_squared = line_direction.norm_squared();
+        if line_length_squared == 0.0 {
+            // The line segment has zero length, return the start point.
+            return self.start;
+        }
+
+        let t = point_direction.dot(&line_direction) / line_length_squared;
+        if t < 0.0 {
+            // The point is closest to the start of the segment.
+            return self.start;
+        } else if t > 1.0 {
+            // The point is closest to the end of the segment.
+            return self.end;
+        }
+
+        // The point is closest to a point on the segment.
+        self.start + t * line_direction
+    }
+
+    // return the closest point on the segment if it falls on him
+    pub fn orthogonal_projection_point_on_segment(
+        &self,
+        point: &Point2<f64>,
+    ) -> Option<Point2<f64>> {
+        let line_direction = self.end - self.start;
+        let point_direction = *point - self.start;
+
+        let line_length_squared = line_direction.norm_squared();
+        if line_length_squared == 0.0 {
+            // The line segment has zero length, return the start point.
+            return None;
+        }
+
+        let t = point_direction.dot(&line_direction) / line_length_squared;
+        if t < 0. || t > 1. {
+            return None;
+        } // The point don't fall on the segment.
+
+        // The point is closest to a point on the segment.
+        Some(self.start + t * line_direction)
+    }
+
+    // takes a number between 0 and 1 and return the point along the line situated at the specified dist
```

Comment:
I don't understand the utility of this function, can you provide me some interesting examples ?

Reply:
Yes, for example, if we want a bot to move along a line at a sinus speed, we can call the sinus function (with time as x) then clamp it between 0 and 1 and then call this function to have the position along the line.
For another example if we want to have the middle of the line we call it with 0.5 in arg, or to have the the first quarter position we call it with 0.25
Hope that helped to understand